megatron/gradient_noise_scale/gradient_noise_scale.py

```diff
@@ -0,0 +1,159 @@
+    batch size (Bsmall).
+
+    when we have those:
+        Bsimple ≈ (tr(Σ) / |G|^2)
+
+    tr(Σ) can be approximated by:
+        tr(Σ) ≈ (1 / ((1/Bsmall) - (1/Bbig))) * (|Gsmall|^2 - |Gbig|^2)
+
+    and |G|^2 by:
+        |G|^2 ≈ (1 / (Bbig - Bsmall)) * (Bbig*|Gbig|^2 - Bsmall*|Gsmall|^2)
+
+    - With multi-gpu training, we can do this by taking the gradients of the microbatch_size_per_gpu for Bsmall,
+    and the gradients of the entire batch for Bbig.
+    - Alternatively, we can just take Bsmall as a single batch, and Bbig as several sequential batches in a row.
+    This is the option we've opted for in this implementation because a) it's easier to implement and b) also works in
+    single-gpu environments. Unfortunately it does come with some memory overhead.
+    """
+
+    def __init__(self, model, batch_size_small, n_batches=10, beta=0.99, cpu_offload=False, args=None, mpu=None):
+        self.batch_size_small = batch_size_small
+        self.batch_size_large = batch_size_small * n_batches
+        self.n_batches = n_batches
+        self.beta = beta
+        self.model = model
+        self.buffer = []
+        self.ema_scale = None
+        self.ema_noise = None
+        self.noise_scale = None
+        self.n_updates = 0
+        self.cpu_offload = cpu_offload
+        self.model.store_gradients = True
+        self.model.store_gradients_cpu = cpu_offload
+        self.args = args
+        self.mpu = mpu
+
+    def flatten_grads(self):
+        grads = []
+        assert hasattr(self.model, 'stored_gradients'), "You might need to update DeeperSpeed"
+        if self.model.stored_gradients is not None:
+            for g in self.model.stored_gradients:
+                if g is not None and not g.isnan().any() and not g.isinf().any():
+                    g = g.flatten().view(-1, 1)
+                    if self.cpu_offload:
+                        g = g.cpu()
+                    grads.append(g)
+                else:
+                    return None
+            if not grads:
+                return None
+            return torch.cat(grads)
+
+    def _sync_overflow(self, is_overflow):
+        if self.args.pipe_parallel_size > 1:
+            # Since each model parallel GPU carries only part of the model,
+            # make sure overflow flag is synced across all the pipe parallel GPUs
+            overflow_gpu = torch.cuda.ByteTensor([is_overflow])
+            torch.distributed.all_reduce(overflow_gpu,
+                                         op=torch.distributed.ReduceOp.MAX,
+                                         group=self.mpu.get_pipe_parallel_group())
+            overflow = overflow_gpu[0].item()
+        else:
+            overflow = is_overflow
+        return overflow
+
+    def _update(self):
+
+        grad = self.flatten_grads()
+        is_overflow = self._sync_overflow(grad is None)
+        if is_overflow:
+            return
+        self.buffer.append(grad)
```

Comment:
Isn't this super expensive? Why not do += and then a division

Reply:
Ah yep well spotted. Updated version works much better. Thanks!